apps/desktop/desktop_native/core/src/ipc/server.rs

```diff
@@ -0,0 +1,190 @@
+use std::error::Error;
+
+use futures::{Sink, SinkExt, TryFutureExt};
+
+use anyhow::Result;
+use interprocess::local_socket::{tokio::prelude::*, GenericFilePath, ListenerOptions};
+use log::{error, info};
+use tokio::{
+    io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt},
+    sync::broadcast::Receiver,
+};
+use tokio_util::sync::CancellationToken;
+
+#[derive(Debug)]
+pub struct Message {
+    pub client_id: u32,
+    pub kind: MessageType,
+    pub message: String,
+}
+
+#[derive(Debug)]
+pub enum MessageType {
+    Connected,
+    Disconnected,
+    Message,
+}
+
+pub struct Server {
+    cancel_token: CancellationToken,
+    tx_connections: tokio::sync::broadcast::Sender<String>,
+}
+
+impl Server {
+    pub fn start<T>(name: &str, tx: T) -> Result<Self, Box<dyn Error>>
+    where
+        T: Sink<Message> + Unpin + Send + Clone + 'static,
+        <T as Sink<Message>>::Error: std::error::Error + 'static,
+    {
+        let path = super::path(name);
+
+        // If the unix socket file already exists, we get an error when trying to bind to it. So we remove it first.
+        // Any processes that were using the old socket should remain connected to it but any new connections will use the new socket.
+        if !cfg!(windows) {
+            let _ = std::fs::remove_file(&path);
+        }
+
+        let name = path.as_os_str().to_fs_name::<GenericFilePath>()?;
+        let opts = ListenerOptions::new().name(name);
+        let listener = opts.create_tokio()?;
+
+        let (tx_connections, rx_connections) = tokio::sync::broadcast::channel::<String>(32);
+
+        let cancel_token = CancellationToken::new();
+        let cancel_token2 = cancel_token.clone();
```

Comment:
suggestion: rename to clarify the purpose/how these will be used

```suggestion
        let upstream_cancel_token = CancellationToken::new();
        let downstream_cancel_token2 = cancel_token.clone();
```
or
```suggestion
        let returned_cancel_token = CancellationToken::new();
        let internal_cancel_token = cancel_token.clone();
```

Reply:
By the way, aren't cancel tokens usually supplied as a parameter?

Reply:
Usually yes, you'd initialize the tokens at the root and then pass them down. In this case I decided to initialize them here because I wasn't sure of a good way to pass them through the NAPI layer, so instead I just exposed a `stop` function, and also implemented `Drop` just in case.